src/shell/autoexec.cpp

```diff
@@ -0,0 +1,675 @@
+		default: assert(false);
+		}
+
+		prints_variables  = (source == Source::Variable);
+		prints_first_line = false;
+
+		for (const auto& autoexec_object : list) {
+			add_string(autoexec_object->GetContent());
+		}
+	}
+}
+
+static void maybe_refresh_blob_dos()
+{
+	// Determine if we have anything to do
+
+	if (shutdown_requested) {
+		// No need to refresh anything during the shutdown
+		return;
+	}
+
+	const auto code_page = UTF8_GetCodePage();
+	if (is_blob_registered && (code_page == last_code_page)) {
+		// No need to refresh anything if code page is the same
+		return;
+	}
+
+	// Regenerate AUTOEXEC.BAT
+
+	std::string blob_dos = {};
+	UTF8_RenderForDos(autoexec_blob_utf8, blob_dos, code_page);
+	last_code_page = code_page;
+
+	// Register/refresh the file
+
+	const auto blob_bin = std::vector<uint8_t>(blob_dos.begin(), blob_dos.end());
+	if (is_blob_registered) {
+		VFILE_Update(autoexec_file_name.c_str(), blob_bin);
+	} else {
+		VFILE_Register(autoexec_file_name.c_str(), blob_bin);
+		is_blob_registered = true;
+	}
+}
+
+// ***************************************************************************
+// AutoExecObject class implementation
+// ***************************************************************************
+
+void AutoExecObject::SetString(const std::string& content, const Source source)
+{
+	assert(source != Source::Variable); // use SetVariable for this source
+
+	ReInstall(autoexec_objects[source]);
+
+	string_content = content;
+}
+
+void AutoExecObject::SetVariable(const std::string& name, const std::string& value)
+{
+#if C_DEBUG
+	auto assert_printable_ascii = [](const std::string& in) {
+		for (const auto character : in) {
+			if (!is_printable_ascii(character)) {
+				E_Exit("AUTOEXEC: Not a printable ASCII");
+			}
+		}
+	};
+
+	assert_printable_ascii(name);
+	assert_printable_ascii(value);
+#endif
+	if (name == variable_name && value == variable_value) {
+		// Same variable, same value - nothing to do
+		return;
+	}
+
+	ReInstall(autoexec_objects[Source::Variable]);
+
+	const auto old_name = variable_name;
+
+	variable_name  = name;
+	variable_value = value;
+	upcase(variable_name);
+
+	// If shell already exists, refresh variable content
+	if (first_shell) {
+		if (!old_name.empty() && variable_name != old_name) {
+			first_shell->SetEnv(old_name.c_str(), "");
+		}
+		first_shell->SetEnv(variable_name.c_str(), variable_value.c_str());
+	}
+}
+
+std::string AutoExecObject::GetContent() const
+{
+	// Handle environment variables
+
+	if (!variable_name.empty()) {
+		const std::string set = "@SET ";
+		return set + variable_name + "=" + variable_value;
+	}
+
+	// Handle string content
+
+	return string_content;
+}
+
+void AutoExecObject::ReInstall(std::list<AutoExecObject*>& list)
+{
+	if (std::find(list.begin(), list.end(), this) == list.end()) {
+		// Remove from other lists
+		UnInstall();
+	}
+
+	if (!is_installed) {
+		list.push_back(this);
+		is_installed = true;
+	}
+}
+
+void AutoExecObject::UnInstall()
```

Comment:
Minor: `Uninstall` (it's a single word, not "un install" or "un-install")

Reply:
Ok, changed.